samples/discoverNatBehavior.c

```diff
@@ -1,109 +1,117 @@
 #include <com/amazonaws/kinesis/video/webrtcclient/Include.h>
 
-#define NETWORK_INTERFACE_NAME_PARAM "-i"
-#define STUN_HOSTNAME_PARAM          "-s"
-#define DEFAULT_STUN_HOST            "stun:stun.sipgate.net:3478"
+#define NETWORK_INTERFACE_NAME_PARAM   "-i"
+#define STUN_HOSTNAME_PARAM            "-s"
+#define MAX_LEN_NETWORK_INTERFACE_NAME 15
 
 BOOL filterFunc(UINT64 data, PCHAR name)
 {
     CHAR* desiredInterface = (CHAR*) data;
-    if (desiredInterface == NULL || STRCMP(name, desiredInterface) == 0) {
+    if (desiredInterface == NULL || STRNCMP(name, desiredInterface, STRNLEN(name, MAX_LEN_NETWORK_INTERFACE_NAME)) == 0) {
         return TRUE;
     }
     return FALSE;
 }
 
 INT32 main(INT32 argc, CHAR** argv)
 {
-    PCHAR interfaceName = NULL, stunHostname = NULL;
-    printf("Usage: ./discoverNatBehavior -i network-interface-name -s stun-hostname\n");
+    PCHAR interfaceName = NULL, stunHostname = NULL, pRegion = NULL, pHostnamePostfix = NULL;
     NAT_BEHAVIOR mappingBehavior = NAT_BEHAVIOR_NONE, filteringBehavior = NAT_BEHAVIOR_NONE;
     INT32 i;
     UINT32 logLevel = LOG_LEVEL_DEBUG;
     STATUS status = STATUS_SUCCESS;
     CHAR* logLevelStr = NULL;
 
+    if ((logLevelStr = GETENV(DEBUG_LOG_LEVEL_ENV_VAR)) != NULL) {
+        status = STRTOUI32(logLevelStr, NULL, 10, &logLevel);
+        if (STATUS_FAILED(status)) {
+            DLOGE("Failed to parse log level with error 0x%08x", status);
+            exit(1);
+        }
+        logLevel = MIN(MAX(logLevel, LOG_LEVEL_VERBOSE), LOG_LEVEL_SILENT);
+    }
+    SET_LOGGER_LOG_LEVEL(logLevel);
+
+    DLOGI("Usage: ./discoverNatBehavior -i network-interface-name -s stun-hostname");
+
+    if ((pRegion = GETENV(DEFAULT_REGION_ENV_VAR)) == NULL) {
+        pRegion = DEFAULT_AWS_REGION;
+    }
+    pHostnamePostfix = KINESIS_VIDEO_STUN_URL_POSTFIX;
+    // If region is in CN, add CN region uri postfix
+    if (STRSTR(pRegion, "cn-")) {
+        pHostnamePostfix = KINESIS_VIDEO_STUN_URL_POSTFIX_CN;
+    }
+    stunHostname = (PCHAR) MEMALLOC((MAX_ICE_CONFIG_URI_LEN + 1) * SIZEOF(CHAR));
+    SNPRINTF(stunHostname, MAX_ICE_CONFIG_URI_LEN + 1, KINESIS_VIDEO_STUN_URL, pRegion, pHostnamePostfix);
+
     for (i = 1; i < argc; ++i) {
-        if (STRCMP(argv[i], NETWORK_INTERFACE_NAME_PARAM) == 0) {
+        if (STRNCMP(argv[i], NETWORK_INTERFACE_NAME_PARAM, STRNLEN(argv[i], MAX_LEN_NETWORK_INTERFACE_NAME)) == 0) {
             interfaceName = argv[++i];
             i++;
```

Comment:
This doesn't seem like a very good use for a for loop why are we incrementing the loop variable twice in the if and the else?  Is that actually intended?  There is already a for loop so the loop variable will be incremented.  Seems like this would be much cleaner with a while loop?

Reply:
Removed the extra increment and tested that the loop works fine without it:
```
niyatim@88665a374cf5 build % ./samples/discoverNatBehavior -i en0 -s stun:stun.kinesisvideo.us-east-1.amazonaws.com:443
2024-02-28 17:16:15.404 INFO    main(): Usage: ./discoverNatBehavior -i network-interface-name -s stun-hostname
2024-02-28 17:16:15.405 INFO    main(): Using stun host: stun:stun.kinesisvideo.us-east-1.amazonaws.com:443, local network interface en0.
2024-02-28 17:16:15.405 INFO    initKvsWebRtc(): Initializing WebRTC library...
2024-02-28 17:16:15.409 INFO    initKvsWebRtc(): SDK version: 3ded8ed68e1464555c2a786e6b0cc690cfcfd506
2024-02-28 17:16:15.410 INFO    initKvsWebRtc(): KVS WebRtc library using thread pool
2024-02-28 17:16:15.410 INFO    createWebRtcClientInstance(): Initialized WebRTC Client instance
2024-02-28 17:16:15.410 INFO    getIpWithHostName(): ICE SERVER Hostname received: stun.kinesisvideo.us-east-1.amazonaws.com
2024-02-28 17:16:15.410 WARN    getIpAddrFromDnsHostname(): Received unexpected hostname format: stun.kinesisvideo.us-east-1.amazonaws.com
2024-02-28 17:16:15.410 WARN    getIpWithHostName(): Parsing for address failed for stun.kinesisvideo.us-east-1.amazonaws.com, fallback to getaddrinfo
2024-02-28 17:16:15.415 INFO    resolveStunIceServerIp(): ICE Server address for stun.kinesisvideo.us-east-1.amazonaws.com with getaddrinfo: 3.217.8.220
2024-02-28 17:16:15.415 PROFILE resolveStunIceServerIp(): [STUN DNS resolution time taken] Time taken: 5 ms
2024-02-28 17:16:15.415 DEBUG   resolveStunIceServerIp(): Exiting from stun server IP resolution thread
2024-02-28 17:16:15.415 PROFILE parseIceServer(): ICE Server address for stun.kinesisvideo.us-east-1.amazonaws.com: 3.217.8.220
2024-02-28 17:16:15.416 DEBUG   createSocketConnection(): create socket with ip: 10.48.230.37:55650. family:1
2024-02-28 17:16:15.416 DEBUG   discoverNatBehavior(): Start NAT mapping behavior test.
2024-02-28 17:16:15.416 DEBUG   discoverNatMappingBehavior(): Running mapping behavior test I. Send binding request
2024-02-28 17:16:15.416 DEBUG   iceUtilsSendStunPacket(): Sending BINDING_REQUEST to ip:3.217.8.220, port:443
2024-02-28 17:16:15.485 ERROR   discoverNatMappingBehavior(): Expect binding response to have other address or changed address attribute
2024-02-28 17:16:15.485 DEBUG   socketConnectionClosed(): Close socket 6
2024-02-28 17:16:15.700 DEBUG   freeSocketConnection(): close socket with ip: 10.48.230.37:55650. family:1
2024-02-28 17:16:15.700 DEBUG   freeSocketConnection(): close socket connected with ip: 0000:0000:0000:0000:0000:0000:0000:0000:0. family:0
2024-02-28 17:16:15.700 DEBUG   createSocketConnection(): create socket with ip: 10.48.230.37:65003. family:1
2024-02-28 17:16:15.700 DEBUG   discoverNatBehavior(): Start NAT filtering behavior test.
2024-02-28 17:16:15.700 DEBUG   discoverNatFilteringBehavior(): Running filtering behavior test I. Send binding request
2024-02-28 17:16:15.700 DEBUG   iceUtilsSendStunPacket(): Sending BINDING_REQUEST to ip:3.217.8.220, port:443
2024-02-28 17:16:15.766 DEBUG   discoverNatFilteringBehavior(): Running filtering behavior test II. Send binding request with change ip and change port flag
2024-02-28 17:16:15.767 DEBUG   iceUtilsSendStunPacket(): Sending BINDING_REQUEST to ip:3.217.8.220, port:443
2024-02-28 17:16:15.831 DEBUG   socketConnectionClosed(): Close socket 6
2024-02-28 17:16:15.831 DEBUG   freeSocketConnection(): close socket with ip: 10.48.230.37:65003. family:1
2024-02-28 17:16:15.831 DEBUG   freeSocketConnection(): close socket connected with ip: 0000:0000:0000:0000:0000:0000:0000:0000:0. family:0
2024-02-28 17:16:15.832 INFO    main(): Detected NAT mapping behavior NONE
2024-02-28 17:16:15.832 INFO    main(): Failed to detect NAT mapping behavior
2024-02-28 17:16:15.832 INFO    main(): Detected NAT filtering behavior ENDPOINT_INDEPENDENT
2024-02-28 17:16:15.832 INFO    main(): Host's NAT allows to receive UDP packet from any external address. STUN is usable.
2024-02-28 17:16:15.832 DEBUG   cleanupWebRtcClientInstance(): Releasing webrtc client context instance from cleanupWebRtcClientInstance
2024-02-28 17:16:15.832 INFO    cleanupWebRtcClientInstance(): Destroyed STUN IP object
2024-02-28 17:16:15.832 INFO    cleanupWebRtcClientInstance(): Destroyed WebRtc client context
2024-02-28 17:16:15.837 INFO    deinitKvsWebRtc(): Destroyed threadpool
niyatim@88665a374cf5 build % 
```
Also, moved the parsing logic before the choosing the server on the basis of the region env var. Use SAFE_MEMFREE instead of MEMFREE